### PR TITLE
Added Currency, DateTime, and PhoneNumber Components and Editors

### PR DESCRIPTION
**Pull request by malvin-softheon**
Added Currency, DateTime, and PhoneNumber Components and Editors

### CODE REVIEW

File: projects/forge-lib/src/lib/shared/form-components/concrete/currency/currency.component.spec.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    .compileComponents();
+  }));
+
+  beforeEach(() => {
+    fixture = TestBed.createComponent(CurrencyComponent);
+    component = fixture.componentInstance;
+    fixture.detectChanges();
+  });
+
+  it('should create', () => {
```
Comment (alamanna) [inline projects/forge-lib/src/lib/shared/form-components/concrete/currency/currency.component.spec.ts]: please unit test the functionality of the component to meet 80% code coverage

File: projects/forge-lib/src/lib/shared/form-editor-components/concrete/currency-editor/data/currency-editor-data.component.html [hunk 2]
```diff
@@ -0,0 +1,4 @@
+<div sws-form input-active full-width>
+  <input [(ngModel)]="this.component.data.inputMask" sws-form__input type="'text'" [name]="'mask_input'" onkeyup="this.setAttribute('value', this.value);">
+  <label sws-form__label [for]="'label_input'">Input Mask</label>
```
Comment (pcohen-softheon) [inline projects/forge-lib/src/lib/shared/form-editor-components/concrete/currency-editor/data/currency-editor-data.component.html]: language file?

File: projects/forge-lib/src/lib/shared/form-editor-components/concrete/currency-editor/display/currency-editor-display.component.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import { Component, OnInit } from '@angular/core';
+import { CurrencyComponent } from '../../../../form-components/concrete/currency/currency.component';
+import { FormEditorDisplayComponent } from '../../../abstract/form-editor/display/form-editor-display-component';
+
+@Component({
+  selector: 'forge-renderer-currency-editor-display',
```
Comment (pcohen-softheon) [inline projects/forge-lib/src/lib/shared/form-editor-components/concrete/currency-editor/display/currency-editor-display.component.ts]: no comments

File: projects/forge-lib/src/lib/shared/form-editor-components/concrete/currency-editor/validation/currency-editor-validation.component.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import { Component, OnInit } from '@angular/core';
+import { FormEditorDisplayComponent } from '../../../abstract/form-editor/display/form-editor-display-component';
+import { CurrencyComponent } from '../../../../form-components/concrete/currency/currency.component';
+
+@Component({
+  selector: 'forge-renderer-currency-editor-validation',
+  templateUrl: './currency-editor-validation.component.html',
+  styleUrls: ['./currency-editor-validation.component.scss']
```
Comment (pcohen-softheon) [inline projects/forge-lib/src/lib/shared/form-editor-components/concrete/currency-editor/validation/currency-editor-validation.component.ts]: same

File: projects/forge-lib/src/lib/shared/form-editor-components/concrete/date-time-editor/data/date-time-editor-data.component.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+import { FormEditorDisplayComponent } from '../../../abstract/form-editor/display/form-editor-display-component';
+
+@Component({
+  selector: 'forge-renderer-date-time-editor-data',
+  templateUrl: './date-time-editor-data.component.html',
+  styleUrls: ['./date-time-editor-data.component.scss']
+})
+export class DateTimeEditorDataComponent extends FormEditorDisplayComponent implements OnInit {
+
+  public component: DateTimeComponent;
```
Comment (pcohen-softheon) [inline projects/forge-lib/src/lib/shared/form-editor-components/concrete/date-time-editor/data/date-time-editor-data.component.ts]: same

File: projects/forge-lib/src/lib/shared/form-editor-components/concrete/date-time-editor/display/date-time-editor-display.component.html [hunk 6]
```diff
@@ -0,0 +1,15 @@
+<div sws-form input-active full-width>
+  <input [(ngModel)]="this.component.display.placeholder" sws-form__input type="'text'" [name]="'placeholder_input'"
+      onkeyup="this.setAttribute('value', this.value);">
+  <label sws-form__label [for]="'label_input'">Placeholder</label>
```
Comment (pcohen-softheon) [inline projects/forge-lib/src/lib/shared/form-editor-components/concrete/date-time-editor/display/date-time-editor-display.component.html]: language file

File: projects/forge-lib/src/lib/shared/form-components/concrete/phone-number/phone-number.component.spec.ts [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    .compileComponents();
+  }));
+
+  beforeEach(() => {
+    fixture = TestBed.createComponent(PhoneNumberComponent);
+    component = fixture.componentInstance;
+    fixture.detectChanges();
+  });
+
+  it('should create', () => {
```
Comment (alamanna) [inline projects/forge-lib/src/lib/shared/form-components/concrete/phone-number/phone-number.component.spec.ts]: potentially missing code coverage

File: projects/forge-lib/src/lib/shared/form-editor-components/concrete/currency-editor/validation/currency-editor-validation.component.ts [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+})
+export class CurrencyEditorValidationComponent extends FormEditorDisplayComponent implements OnInit {
+
+  public component: CurrencyComponent;
+
+  ngOnInit() {
```
Comment (alamanna) [inline projects/forge-lib/src/lib/shared/form-editor-components/concrete/currency-editor/validation/currency-editor-validation.component.ts]: access modifier and return type

File: projects/forge-lib/src/lib/shared/form-editor-components/concrete/date-time-editor/data/date-time-editor-data.component.ts [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+
+  ngOnInit() {
```
Comment (alamanna) [inline projects/forge-lib/src/lib/shared/form-editor-components/concrete/date-time-editor/data/date-time-editor-data.component.ts]: same

File: projects/forge-lib/src/lib/shared/form-editor-components/concrete/date-time-editor/display/date-time-editor-display.component.ts [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+
+@Component({
+  selector: 'forge-renderer-date-time-editor-display',
+  templateUrl: './date-time-editor-display.component.html',
+  styleUrls: ['./date-time-editor-display.component.scss']
+})
+export class DateTimeEditorDisplayComponent extends FormEditorDisplayComponent implements OnInit {
+
+  public component: DateTimeComponent;
+
```
Comment (alamanna) [inline projects/forge-lib/src/lib/shared/form-editor-components/concrete/date-time-editor/display/date-time-editor-display.component.ts]: comments and access modifiers and return types

File: projects/forge-lib/src/lib/shared/form-editor-components/concrete/date-time-editor/validation/date-time-editor-validation.component.ts [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+  selector: 'forge-renderer-date-time-editor-validation',
+  templateUrl: './date-time-editor-validation.component.html',
+  styleUrls: ['./date-time-editor-validation.component.scss']
+})
+export class DateTimeEditorValidationComponent extends FormEditorDisplayComponent implements OnInit {
+
+  public component: DateTimeComponent;
+
+  ngOnInit() {
+  }
```
Comment (alamanna) [inline projects/forge-lib/src/lib/shared/form-editor-components/concrete/date-time-editor/validation/date-time-editor-validation.component.ts]: same

File: projects/forge-lib/src/lib/shared/form-editor-components/concrete/phone-number-editor/data/phone-number-editor-data.component.ts [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+  selector: 'forge-renderer-phone-number-editor-data',
+  templateUrl: './phone-number-editor-data.component.html',
+  styleUrls: ['./phone-number-editor-data.component.scss']
+})
+export class PhoneNumberEditorDataComponent extends FormEditorDisplayComponent implements OnInit {
+
+  public component: PhoneNumberComponent;
+
+  ngOnInit() {
+  }
```
Comment (alamanna) [inline projects/forge-lib/src/lib/shared/form-editor-components/concrete/phone-number-editor/data/phone-number-editor-data.component.ts]: same

File: projects/forge-lib/src/lib/shared/form-editor-components/concrete/phone-number-editor/display/phone-number-editor-display.component.ts [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+
+@Component({
+  selector: 'forge-renderer-phone-number-editor-display',
+  templateUrl: './phone-number-editor-display.component.html',
+  styleUrls: ['./phone-number-editor-display.component.scss']
+})
+export class PhoneNumberEditorDisplayComponent extends FormEditorDisplayComponent implements OnInit {
+
+  public component: PhoneNumberComponent;
+
```
Comment (alamanna) [inline projects/forge-lib/src/lib/shared/form-editor-components/concrete/phone-number-editor/display/phone-number-editor-display.component.ts]: samew

File: projects/forge-lib/src/lib/shared/form-editor-components/concrete/phone-number-editor/validation/phone-number-editor-validation.component.ts [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+
+@Component({
+  selector: 'forge-renderer-phone-number-editor-validation',
+  templateUrl: './phone-number-editor-validation.component.html',
+  styleUrls: ['./phone-number-editor-validation.component.scss']
+})
+export class PhoneNumberEditorValidationComponent extends FormEditorDisplayComponent implements OnInit {
+
+  public component: PhoneNumberComponent;
+
```
Comment (alamanna) [inline projects/forge-lib/src/lib/shared/form-editor-components/concrete/phone-number-editor/validation/phone-number-editor-validation.component.ts]: same

File: projects/forge-lib/src/lib/shared/form-editor-components/concrete/currency-editor/data/currency-editor-data.component.html (hunk 2 above)
Comment (malvin-softheon) [reply projects/forge-lib/src/lib/shared/form-editor-components/concrete/currency-editor/data/currency-editor-data.component.html]: language file will be added on later phases

File: projects/forge-lib/src/lib/shared/form-editor-components/concrete/date-time-editor/display/date-time-editor-display.component.html (hunk 6 above)
Comment (malvin-softheon) [reply projects/forge-lib/src/lib/shared/form-editor-components/concrete/date-time-editor/display/date-time-editor-display.component.html]: language file will be added on later phases

**Comment by malvin-softheon** [comment]
Code coverage should be > 80% @alamanna

File: projects/forge-lib/src/lib/shared/directives/phone-number.directive.ts [hunk 15]
```diff
@@ -0,0 +1,63 @@
+import { Directive, Input, HostListener, ElementRef } from '@angular/core';
+
+@Directive({
+    selector: '[forgePhoneNumber]'
+})
+export class PhoneNumberDirective {
+    constructor(private el: ElementRef) { }
+
```
Comment (pcohen-softheon) [inline projects/forge-lib/src/lib/shared/directives/phone-number.directive.ts]: missing comment

File: projects/forge-lib/src/lib/shared/form-components/concrete/currency/currency.component.html [hunk 16]
```diff
@@ -0,0 +1,30 @@
+<div flex-container-row>
+  <button [id]="'prefix_' + id" class="prefix-suffix" sws-button-simple--jet m-t-25 m-r-0
+    *ngIf="display.prefix">{{display.prefix}}</button>
+  <div sws-form input-active full-width>
+    <input [id]="'input_' + id" forgeNumber [requireDecimal]="true" [decimalPlaces]="2" [(ngModel)]="value"
+      sws-form__input type="text" #name="ngModel" [name]="'input_' + id" [disabled]="display.disabled"
+      [required]="validation.required" onkeyup="this.setAttribute('value', this.value);">
```
Comment (pcohen-softheon) [inline projects/forge-lib/src/lib/shared/form-components/concrete/currency/currency.component.html]: you shouldnt need the "this" context in the HTML

File: projects/forge-lib/src/lib/modules/form-builder/components/field-editor/field-editor.component.ts [hunk 17]
```diff
@@ -221,6 +221,34 @@ export class FieldEditorComponent implements OnInit {
         this.showAbstractDisplay = true;
         this.showAbstractData = true;
         this.showAbstractValidation = true;
+        break;
+      }
+      case ComponentTypes.Currency: {
```
Comment (alamanna) [inline projects/forge-lib/src/lib/modules/form-builder/components/field-editor/field-editor.component.ts]: Does the the code coverage meet the 80% rule here with updates?

File: projects/forge-lib/src/lib/shared/form-components/concrete/date-time/date-time-display.ts [hunk 18]
```diff
@@ -0,0 +1,18 @@
+import { FormDisplay } from '../../abstract/form-display';
+
```
Comment (pcohen-softheon) [inline projects/forge-lib/src/lib/shared/form-components/concrete/date-time/date-time-display.ts]: class comment

File: projects/forge-lib/src/lib/shared/form-components/concrete/date-time/date-time.component.ts [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+   * api attributes
+   */
+  public api: FormApi = new FormApi();
+
+  /**
+   * Value model
+   */
+  public value: string;
+
+
```
Comment (pcohen-softheon) [inline projects/forge-lib/src/lib/shared/form-components/concrete/date-time/date-time.component.ts]: nit: extra line

File: projects/forge-lib/src/lib/shared/form-components/concrete/date-time/date-time-display.ts [hunk 20]
```diff
@@ -0,0 +1,18 @@
...
+export class DateTimeDisplay extends FormDisplay {
```
Comment (alamanna) [inline projects/forge-lib/src/lib/shared/form-components/concrete/date-time/date-time-display.ts]: missing class comments

File: projects/forge-lib/src/lib/shared/form-editor-components/concrete/currency-editor/data/currency-editor-data.component.ts [hunk 21]
```diff
@@ -0,0 +1,18 @@
...
+import { FormEditorDisplayComponent } from '../../../abstract/form-editor/display/form-editor-display-component';
+
+@Component({
+  selector: 'forge-renderer-currency-editor-data',
+  templateUrl: './currency-editor-data.component.html',
+  styleUrls: ['./currency-editor-data.component.scss']
+})
+export class CurrencyEditorDataComponent  extends FormEditorDisplayComponent implements OnInit {
+
+  public component: CurrencyComponent;
```
Comment (alamanna) [inline projects/forge-lib/src/lib/shared/form-editor-components/concrete/currency-editor/data/currency-editor-data.component.ts]: missing comments

File: projects/forge-lib/src/lib/shared/form-editor-components/concrete/currency-editor/display/currency-editor-display.component.ts (hunk 3 above)
Comment (alamanna) [reply projects/forge-lib/src/lib/shared/form-editor-components/concrete/currency-editor/display/currency-editor-display.component.ts]: doesn't look like the file got updated with comments here

File: projects/forge-lib/src/lib/shared/form-editor-components/concrete/currency-editor/validation/currency-editor-validation.component.ts (hunk 4 above)
Comment (alamanna) [reply projects/forge-lib/src/lib/shared/form-editor-components/concrete/currency-editor/validation/currency-editor-validation.component.ts]: same

File: projects/forge-lib/src/lib/shared/form-editor-components/concrete/date-time-editor/data/date-time-editor-data.component.ts (hunk 5 above)
Comment (alamanna) [reply projects/forge-lib/src/lib/shared/form-editor-components/concrete/date-time-editor/data/date-time-editor-data.component.ts]: same as above

File: projects/forge-lib/src/lib/shared/form-components/concrete/currency/currency.component.html (hunk 16 above)
Comment (malvin-softheon) [reply projects/forge-lib/src/lib/shared/form-components/concrete/currency/currency.component.html]: I believe this is used when we have a directive to manipulate the DOM. I was following the structure of other components that were already developed and they used the same onkeyup="this.setAttribute...."

File: projects/forge-lib/src/lib/modules/form-builder/components/field-editor/field-editor.component.ts (hunk 17 above)
Comment (malvin-softheon) [reply projects/forge-lib/src/lib/modules/form-builder/components/field-editor/field-editor.component.ts]: Now it should be > 80% coverage, after my most recent push.

File: projects/forge-lib/src/lib/shared/directives/phone-number.directive.ts [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+    @Input() inputMask: string;
+
+    @HostListener('keyup', ['$event']) onkeyUp(_) {
```
Comment (edriscoll-softheon) [inline projects/forge-lib/src/lib/shared/directives/phone-number.directive.ts]: Missing comment

File: projects/forge-lib/src/lib/shared/directives/phone-number.directive.ts [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+        }
+
+        // Unformat phone number if already formatted but not complete numbers
+        if ((value.replace(/[^0-9]/g, "").length < this.inputMask.split("x").length - 1)
+            && (value.replace(/[0-9]/g, "").length != 0)) {
+            this.el.nativeElement.value = this.el.nativeElement.value.replace(/[^0-9]/g, "");
+        }
+    }
+
+    @HostListener('keydown', ['$event']) onKeyDown(event) {
```
Comment (edriscoll-softheon) [inline projects/forge-lib/src/lib/shared/directives/phone-number.directive.ts]: missing comment

File: projects/forge-lib/src/lib/shared/validators/dateValidator.ts [hunk 24]
```diff
@@ -0,0 +1,29 @@
...
+      (val.includes('d') || val.includes('dd')) && (val.includes('yyyy')));
+    var valid = false;
+    if (val.match(format)) {
+      valid = (val.match(format).length == 1) && (required);
+    }
+    return !valid ? { 'invalidDateFormat': { value: control.value } } : null;
+  };
+}
+
+@Directive({
```
Comment (alamanna) [inline projects/forge-lib/src/lib/shared/validators/dateValidator.ts]: nit comment

File: projects/forge-lib/src/lib/shared/validators/phoneMaskValidator.ts [hunk 25]
```diff
@@ -0,0 +1,32 @@
...
+    var numericCharsCount = 0;
+    if(val.split('x')) {
+        numericCharsCount = val.split('x').length - 1;
+    }
+    const valid = ((specialCharsCount + numericCharsCount) == (val.length));
+    return !valid ? { 'invalidPhoneMask': { value: control.value } } : null;
+  };
+}
+
+@Directive({
```
Comment (alamanna) [inline projects/forge-lib/src/lib/shared/validators/phoneMaskValidator.ts]: nit: comment

**Pull request merged**